File: mobile_app/lib/src/test_engine.dart
import 'dart:math';

import 'models.dart';
import 'storage.dart';

class TestEngine {
  TestEngine({
    required this.version,
    required this.questions,
    required this.storage,
    SavedProgress? initialProgress,
  }) : _answers = Map<int, String>.from(initialProgress?.answers ?? {}),
       _currentIndex = initialProgress?.currentIndex ?? 0,
       _startTime =
           initialProgress?.startTime ?? DateTime.now().millisecondsSinceEpoch;

  final VersionId version;
  final List<QuestionModel> questions;
  final AppStorage storage;

  final Map<int, String> _answers;
  int _currentIndex;
  final int _startTime;

  int get currentIndex => _currentIndex;
  int get totalQuestions => questions.length;
  QuestionModel get currentQuestion => questions[_currentIndex];
  String? get currentAnswer => _answers[currentQuestion.id];
  Map<int, String> get answers => Map<int, String>.unmodifiable(_answers);
  int get progressPercent =>
      ((_answers.length / max(questions.length, 1)) * 100).round();
  bool get isComplete => _answers.length == questions.length;

  Future<void> ensureProgress() => _persist();

  Future<void> answer(String choice) async {
    _answers[currentQuestion.id] = choice;
    await _persist();
  }

  Future<void> next() async {
    if (_currentIndex < questions.length - 1) {
      _currentIndex += 1;
      await _persist();
    }
  }

  Future<void> previous() async {
    if (_currentIndex > 0) {
      _currentIndex -= 1;
      await _persist();
    }
  }

  Future<void> jumpTo(int index) async {
    if (index >= 0 && index < questions.length) {
      _currentIndex = index;
      await _persist();
    }
  }

  Future<void> clearProgress() => storage.clearProgress(version);

  TestResultModel calculateResult() {
    final scores = <String, int>{
      'E': 0,
      'I': 0,
      'S': 0,
      'N': 0,
      'T': 0,
      'F': 0,
      'J': 0,
      'P': 0,
    };

    for (final question in questions) {
      final choice = _answers[question.id];
      if (choice == null) continue;
      final selected = question.options.firstWhere(
        (item) => item.label == choice,
      );
      scores[selected.value] = (scores[selected.value] ?? 0) + 1;
    }

    final resultType = [
      scores['E']! >= scores['I']! ? 'E' : 'I',
      scores['S']! >= scores['N']! ? 'S' : 'N',
      scores['T']! >= scores['F']! ? 'T' : 'F',
      scores['J']! >= scores['P']! ? 'J' : 'P',
    ].join();

    final threshold = switch (version) {
      VersionId.quick => 1,
      VersionId.standard => 3,
      VersionId.full => 5,
    };

    String dimensionValue(String a, String b) {
      final left = scores[a]!;
      final right = scores[b]!;
      if ((left - right).abs() <= threshold) return 'X';
      return left > right ? a : b;
    }

    return TestResultModel(
      id: '${DateTime.now().millisecondsSinceEpoch}-${Random().nextInt(999999)}',
      timestamp: DateTime.now().millisecondsSinceEpoch,
      version: version,
      scores: TestScores(
        e: scores['E']!,
        i: scores['I']!,
        s: scores['S']!,
        n: scores['N']!,
        t: scores['T']!,
        f: scores['F']!,
        j: scores['J']!,
        p: scores['P']!,
      ),
      resultType: resultType,
      dimensions: {
        'EI': dimensionValue('E', 'I'),
        'SN': dimensionValue('S', 'N'),
        'TF': dimensionValue('T', 'F'),
        'JP': dimensionValue('J', 'P'),
      },
    );
  }

  Future<void> _persist() async {
    await storage.saveProgress(
      SavedProgress(
        version: version,
        answers: _answers,
        currentIndex: _currentIndex,
        startTime: _startTime,
        lastUpdate: DateTime.now().millisecondsSinceEpoch,
      ),
    );
  }
}

File: apps/web/src/lib/TestEngine.ts
import { Question, SupportedLocale, TestResult, VersionId } from '../types';
import { LocalStorageManager } from './LocalStorageManager';
import { getLocalizedQuestions } from './localeData';

export class TestEngine {
  private questions: Question[];
  private questionMap: Map<number, Question>;
  private currentIndex: number;
  private answers: Map<number, 'A' | 'B'>;
  private version: VersionId;
  private startTime: number;

  constructor(version: VersionId, locale: SupportedLocale = LocalStorageManager.getLanguage()) {
    this.version = version;
    this.questions = getLocalizedQuestions(locale)[version];
    this.questionMap = new Map(this.questions.map((question) => [question.id, question]));
    this.currentIndex = 0;
    this.answers = new Map();
    this.startTime = Date.now();
  }

  public getQuestions(): Question[] {
    return this.questions;
  }

  public getVersion(): VersionId {
    return this.version;
  }

  public answerQuestion(index: number, choice: 'A' | 'B'): void {
    const question = this.questions[index];
    if (question) {
      this.answers.set(question.id, choice);
      this.saveProgress();
    }
  }

  public getCurrentQuestion(): Question | null {
    return this.questions[this.currentIndex] || null;
  }

  public nextQuestion(): boolean {
    if (this.currentIndex < this.questions.length - 1) {
      this.currentIndex++;
      this.saveProgress();
      return true;
    }

    return false;
  }

  public previousQuestion(): boolean {
    if (this.currentIndex > 0) {
      this.currentIndex--;
      this.saveProgress();
      return true;
    }

    return false;
  }

  public jumpToQuestion(index: number): boolean {
    if (index >= 0 && index < this.questions.length) {
      this.currentIndex = index;
      this.saveProgress();
      return true;
    }

    return false;
  }

  public getProgress(): number {
    return Math.round((this.answers.size / this.questions.length) * 100);
  }

  public getAnswer(questionId: number): 'A' | 'B' | undefined {
    return this.answers.get(questionId);
  }

  public isComplete(): boolean {
    return this.answers.size === this.questions.length;
  }

  private generateUUID(): string {
    if (typeof crypto !== 'undefined' && crypto.randomUUID) {
      return crypto.randomUUID();
    }

    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (token) => {
      const random = (Math.random() * 16) | 0;
      const value = token === 'x' ? random : (random & 0x3) | 0x8;
      return value.toString(16);
    });
  }

  public calculateScores(): TestResult {
    const scores = {
      E: 0,
      I: 0,
      S: 0,
      N: 0,
      T: 0,
      F: 0,
      J: 0,
      P: 0,
    };

    this.answers.forEach((choice, id) => {
      const question = this.questionMap.get(id);
      if (!question) return;

      const selectedOption = question.options.find((option) => option.label === choice);
      if (!selectedOption) return;

      const value = selectedOption.value;
      if (scores[value] !== undefined) {
        scores[value]++;
      }
    });

    const resultType = [
      scores.E >= scores.I ? 'E' : 'I',
      scores.S >= scores.N ? 'S' : 'N',
      scores.T >= scores.F ? 'T' : 'F',
      scores.J >= scores.P ? 'J' : 'P',
    ].join('');

    const thresholds = { quick: 1, standard: 3, full: 5 } as const;
    const threshold = thresholds[this.version];

    const dimensions = {
      EI: Math.abs(scores.E - scores.I) <= threshold ? 'X' : scores.E > scores.I ? 'E' : 'I',
      SN: Math.abs(scores.S - scores.N) <= threshold ? 'X' : scores.S > scores.N ? 'S' : 'N',
      TF: Math.abs(scores.T - scores.F) <= threshold ? 'X' : scores.T > scores.F ? 'T' : 'F',
      JP: Math.abs(scores.J - scores.P) <= threshold ? 'X' : scores.J > scores.P ? 'J' : 'P',
    } as TestResult['dimensions'];

    return {
      id: this.generateUUID(),
      timestamp: Date.now(),
      version: this.version,
      scores,
      resultType,
      dimensions,
    };
  }

  public saveProgress(): void {
    const answersObj: Record<number, 'A' | 'B'> = {};
    this.answers.forEach((value, key) => {
      answersObj[key] = value;
    });

    LocalStorageManager.saveCurrentTest(this.version, {
      version: this.version,
      answers: answersObj,
      currentIndex: this.currentIndex,
      startTime: this.startTime,
      lastUpdate: Date.now(),
    });
  }

  public loadProgress(): void {
    const data = LocalStorageManager.getSavedTest(this.version);
    if (data && data.version === this.version) {
      this.currentIndex = data.currentIndex;
      this.startTime = data.startTime;
      this.answers = new Map();
      Object.entries(data.answers).forEach(([key, value]) => {
        this.answers.set(Number(key), value);
      });
    }
  }
}
